File: javascore/bmv/testsvc/src/main/java/com/iconloop/testsvc/Account.java
/*
 * Copyright 2020 ICONLOOP Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.iconloop.testsvc;

import score.Address;

import java.math.BigInteger;
import java.util.HashMap;
import java.util.Map;

public class Account {
    private static final Map<Address, Account> accounts = new HashMap<>();

    private final Address address;
    private final Map<String, BigInteger> balances = new HashMap<>();

    public static Account newExternalAccount(int seed) {
        var acct = new Account(0, seed);
        accounts.put(acct.getAddress(), acct);
        return acct;
    }

    public static Account newScoreAccount(int seed) {
        var acct = new Account(1, seed);
        accounts.put(acct.getAddress(), acct);
        return acct;
    }

    public static Account getAccount(Address address) {
        return accounts.get(address);
    }

    private Account(int type, int seed) {
        var ba = new byte[Address.LENGTH];
        ba[0] = (byte) type;
        var index = ba.length - 1;
        ba[index--] = (byte) seed;
        ba[index--] = (byte) (seed >> 8);
        ba[index--] = (byte) (seed >> 16);
        ba[index] = (byte) (seed >> 24);
        this.address = new Address(ba);
    }

    public Address getAddress() {
        return address;
    }

    public void addBalance(String symbol, BigInteger value) {
        balances.put(symbol, getBalance(symbol).add(value));
    }

    public void subtractBalance(String symbol, BigInteger value) {
        balances.put(symbol, getBalance(symbol).subtract(value));
    }

    public BigInteger getBalance(String symbol) {
        return balances.getOrDefault(symbol, BigInteger.ZERO);
    }

    public BigInteger getBalance() {
        return getBalance("ICX");
    }

    @Override
    public String toString() {
        return "Account{" +
                "address=" + address +
                ", balances=" + balances +
                '}';
    }
}

File: chain/near/account/account.go
/*
 * Copyright 2021 ICON Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package account

import (
	"crypto/ed25519"
	"crypto/rand"
	"encoding/hex"
	"errors"

	"github.com/btcsuite/btcutil/base58"
)

func GenerateKey() ([]byte, []byte, error) {
	publicKey, privateKey, err := ed25519.GenerateKey(rand.Reader)

	if err != nil {
		return nil, nil, err
	}

	return privateKey.Seed(), publicKey, nil
}
func GeneratePubKeyBySeed(seed []byte) ([]byte, []byte, error) {
	privateKey := ed25519.NewKeyFromSeed(seed)
	publicKey := privateKey.Public().(ed25519.PublicKey)

	if len(publicKey) != 32 {
		return nil, nil, errors.New("public key length is not equal 32")
	}

	return publicKey, privateKey, nil
}

func GeneratePubKeyByBase58(b58Key string) ([]byte, []byte, error) {
	seed := base58.Decode(b58Key)

	if len(seed) == 0 {
		return nil, nil, errors.New("base 58 decode error")
	}

	if len(seed) != 32 {
		return nil, nil, errors.New("seed length is not equal 32")
	}

	return GeneratePubKeyBySeed(seed)
}

func PublicKeyToString(pub []byte) string {
	publicKey := base58.Encode(pub)
	return "ed25519:" + publicKey
}

func KeyToAddress(key []byte) string {
	return hex.EncodeToString(key)
}
